temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+        self, client: Client, id: str, *, run_id: Optional[str] = None
+    ) -> None:
+        """Create workflow handle."""
+        self._client = client
+        self._id = id
+        self._run_id = run_id
+
+    @property
+    def id(self) -> str:
+        """ID for the workflow."""
+        return self._id
+
+    @property
+    def run_id(self) -> Optional[str]:
+        """Run ID used for calls on this handle if present."""
+        return self._run_id
+
+    async def result(
+        self, *, starting_run_id: Optional[str] = SELF_RUN_ID, follow_runs: bool = True
+    ) -> T:
+        """Wait for result of the workflow.
+
+        Args:
+            starting_run_id: Run ID to fetch result for. Defaults to using
+                :py:meth:`run_id`. If set to None or there is no
+                :py:meth:`run_id`, this will get the latest result for the
+                workflow ID.
+            follow_runs: If true (default), workflow runs will be continually
+                fetched, until the most recent one is found. If false, the first
+                result is used.
+
+        Returns:
+            Result of the workflow after being converted by the data converter.
+
+        Raises:
+            Exception: Any failure of the workflow.
+        """
+        if starting_run_id == WorkflowHandle.SELF_RUN_ID:
+            starting_run_id = self._run_id
+        req = temporalio.api.workflowservice.v1.GetWorkflowExecutionHistoryRequest(
+            namespace=self._client.namespace,
+            execution=temporalio.api.common.v1.WorkflowExecution(
+                workflow_id=self._id, run_id=starting_run_id or ""
+            ),
+            wait_new_event=True,
+            history_event_filter_type=temporalio.api.enums.v1.HistoryEventFilterType.HISTORY_EVENT_FILTER_TYPE_CLOSE_EVENT,
+            skip_archival=True,
+        )
+        while True:
+            resp = await self._client.service.get_workflow_execution_history(req)
+            # Continually ask for pages until we get close
+            if len(resp.history.events) == 0:
+                req.next_page_token = resp.next_page_token
+                continue
+            elif len(resp.history.events) != 1:
+                raise RuntimeError(
+                    f"Expected single close event, got {len(resp.history.events)}"
+                )
+            event = resp.history.events[0]
+            if event.HasField("workflow_execution_completed_event_attributes"):
+                complete_attr = event.workflow_execution_completed_event_attributes
+                # Follow execution
+                if follow_runs and complete_attr.new_execution_run_id:
+                    req.execution.run_id = complete_attr.new_execution_run_id
+                    req.next_page_token = b""
+                    continue
+                # Ignoring anything after the first response like TypeScript
+                if not complete_attr.result:
+                    return cast(T, None)
```

Comment:
Do we want to lie here and return None or throw?
In TS we just ignore whatever's returned because we don't have types at runtime.
I think it's reasonable to check the return type at runtime (e.g. with https://typeguard.readthedocs.io/en/latest/)

Reply:
> Do we want to lie here and return None or throw?

Just return, like we would in other dynamic languages

> In TS we just ignore whatever's returned because we don't have types at runtime.

We don't in Python either.

> I think it's reasonable to check the return type at runtime (e.g. with https://typeguard.readthedocs.io/en/latest/)

I don't think that's reasonable at all, especially over a generic type. I think it's unreasonable and not Pythonic at all.

Python type hints are known to be dev-time utilities, not runtime checked in most cases. We don't need to deviate here.